File: DoublyLinkedList.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
        self.prev = None


class DoublyLinkedList:
    def __init__(self, value=None):
        if value is None:
            self.head = None
            self.tail = None
            self.length = 0
            return

        new_node = Node(value)
        self.head = new_node
        self.tail = new_node
        self.length = 1

    def print_list(self):
        if self.head is None:
            print("Empty Linked List")
            return

        temp = self.head
        while temp is not None:
            print(temp.value)
            temp = temp.next

    def append(self, value):
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
            self.length += 1
            return True

        self.tail.next = new_node
        new_node.prev = self.tail
        self.tail = new_node
        self.length += 1
        return True

    def pop(self):
        if self.head is None:
            return None

        temp = self.tail
        if self.length == 1:
            self.head = None
            self.tail = None
            self.length -= 1
            return temp

        self.tail = temp.prev
        self.tail.next = None
        temp.prev = None
        self.length -= 1
        return temp

    def prepend(self, value):
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
            self.length += 1
            return True

        new_node.next = self.head
        self.head.prev = new_node
        self.head = new_node
        self.length += 1
        return True

    def pop_first(self):
        if self.head is None:
            return None

        temp = self.head
        if self.length == 1:
            self.head = None
            self.tail = None
            self.length -= 1
            return temp

        self.head = temp.next
        self.head.prev = None
        temp.next = None
        self.length -= 1
        return temp

    def get(self, index):
        if index < 0 or index >= self.length:
            return None

        if index < self.length / 2:
            temp = self.head
            for _ in range(index):
                temp = temp.next
            return temp

        temp = self.tail
        for _ in range(self.length - 1, index, -1):
            temp = temp.prev
        return temp

    def set_value(self, index, value):
        temp = self.get(index)
        if temp is None:
            return False

        temp.value = value
        return True

    def insert(self, index, value):
        if index < 0 or index > self.length:
            return False

        if index == 0:
            return self.prepend(value)

        if index == self.length:
            return self.append(value)

        new_node = Node(value)
        before = self.get(index - 1)
        after = before.next

        before.next = new_node
        new_node.prev = before
        new_node.next = after
        after.prev = new_node

        self.length += 1
        return True

    def remove(self, index):
        if index < 0 or index >= self.length:
            return None

        if index == 0:
            return self.pop_first()

        if index == self.length - 1:
            return self.pop()

        temp = self.get(index)
        before = temp.prev
        after = temp.next

        before.next = after
        after.prev = before
        temp.next = None
        temp.prev = None

        self.length -= 1
        return temp


myDLL = DoublyLinkedList(4)
myDLL.append(5)
myDLL.append(6)
myDLL.append(7)
myDLL.append(8)
myDLL.append(9)
myDLL.set_value(3, 100)

myDLL.remove(0)
myDLL.remove(0)
myDLL.remove(0)
myDLL.remove(0)

myDLL.print_list()

File: DoublyLinkedList.cpp
#include <iostream>
using namespace std;

class Node {
  public:
    int value;
    Node* next;
    Node* prev;

    Node(int val) {
        value = val;
        next = nullptr;
        prev = nullptr;
    }
};

class DoublyLinkedList {
  private:
    Node* head;
    Node* tail;
    int length;

  public:
    /* Constructor */
    DoublyLinkedList() {
        head = nullptr;
        tail = nullptr;
        length = 0;
    }

    DoublyLinkedList(int value) {
        Node* newNode = new Node(value);
        head = newNode;
        tail = newNode;
        length = 1;
    }

    /* Destructor */
    ~DoublyLinkedList() {
        Node* temp = head;
        while (temp != nullptr) {
            head = head->next;
            delete temp;
            temp = head;
        }
    }

    /* Methods */
    void printList() {
        if (head == nullptr) {
            cout << "Empty Linked List" << endl;
            return;
        }

        Node* currentNode = head;
        while (currentNode != nullptr) {
            cout << currentNode->value << " ";
            currentNode = currentNode->next;
        }
        cout << endl;
    }

    void getHead() {
        if (head != nullptr) {
            cout << "Head: " << head->value << endl;
            return;
        }
        cout << "Head: nullptr." << endl;
    }

    void getTail() {
        if (tail != nullptr) {
            cout << "Tail: " << tail->value << endl;
            return;
        }
        cout << "Tail: nullptr." << endl;
    }

    void getLength() {
        cout << "Length: " << length << endl;
    }

    bool append(int value) {
        Node* newNode = new Node(value);
        if (head == nullptr) {
            head = newNode;
            tail = newNode;
            length++;
            return true;
        }

        tail->next = newNode;
        newNode->prev = tail;
        tail = newNode;
        length++;
        return true;
    }

    void deleteLast() {
        if (head == nullptr) return;
        if (length == 1) {
            delete head;
            head = nullptr;
            tail = nullptr;
            length--;
            return;
        }

        Node* temp = tail;
        tail = tail->prev;
        tail->next = nullptr;
        temp->prev = nullptr;
        length--;
        delete temp;
    }

    bool prepend(int value) {
        Node* newNode = new Node(value);
        if (head == nullptr) {
            head = newNode;
            tail = newNode;
            length++;
            return true;
        }

        newNode->next = head;
        head->prev = newNode;
        head = newNode;
        length++;
        return true;
    }

    void deleteFirst() {
        if (head == nullptr) return;
        if (length == 1) {
            delete head;
            head = nullptr;
            tail = nullptr;
            length--;
            return;
        }

        Node* temp = head;
        head = head->next;
        head->prev = nullptr;
        temp->next = nullptr;
        length--;
        delete temp;
    }

    Node* get(int index) {
        if (index < 0 || index >= length) return nullptr;
        Node* temp = head;
        if (index < length / 2) {
            for (int i = 0; i < index; i++) {
                temp = temp->next;
            }
        } else {
            temp = tail;
            for (int i = length - 1; i > index; i--) {
                temp = temp->prev;
            }
        }

        return temp;
    }

    bool set(int index, int value) {
        Node* temp = get(index);
        if (temp == nullptr) return false;
        temp->value = value;
        return true;
    }

    bool insert(int index, int value) {
        if (index < 0 || index > length) return false;
        if (index == 0) return prepend(value);
        if (index == length) return append(value);

        Node* newNode = new Node(value);
        Node* before = get(index - 1);
        Node* after = before->next;

        before->next = newNode;
        newNode->prev = before;
        newNode->next = after;
        after->prev = newNode;
        length++;
        return true;
    }

    void deleteNode(int index) {
        if (index < 0 || index >= length) return;
        if (index == 0) return deleteFirst();
        if (index == length - 1) return deleteLast();

        Node* temp = get(index);
        Node* before = temp->prev;
        Node* after = temp->next;

        before->next = after;
        after->prev = before;
        temp->next = nullptr;
        temp->prev = nullptr;
        length--;
        delete temp;
    }
};

int main() {
    DoublyLinkedList* myDLL = new DoublyLinkedList(1);
    myDLL->append(2);
    myDLL->append(3);
    myDLL->prepend(99);
    myDLL->deleteNode(0);
    myDLL->deleteNode(0);
    myDLL->deleteNode(0);
    myDLL->deleteNode(0);
    myDLL->deleteNode(0);

    myDLL->printList();

    delete myDLL;
    return 0;
}